KEYCLOAK-1267 Add dedicated SSO timeouts for Remember-Me

thomasdarimont opened the pull request:
Previously remember-me sessions where tied to the SSO max session
timeout which could lead to unexpected early session timeouts.
We now allow SSO timeouts to be configured separately for sessions
with enabled remember-me.
This enables users to opt-in for longer session timeouts.

SSO Session Timeouts for RememberMe can now be configured in the
tokens tab in the realm admin console. Max SSO timeouts for Remember-me
need to be equal or greater than normal max SSO timeouts.

Note that most of the work for this PR was done by @cfsnyder and
is based on his earlier PR https://github.com/keycloak/keycloak/pull/2338.
I just adapted his code to the latest Keycloak API changes.

Signed-off-by: Thomas Darimont thomas.darimont@gmail.com

thomasdarimont committed:
KEYCLOAK-1267 Add dedicated SSO timeouts for Remember-Me

Previosuly remember-me sessions where tied to the SSO max session
timeout which could lead to unexpected early session timeouts.
We now allow SSO timeouts to be configured separately for sessions
with enabled remember-me.
This enables users to opt-in for longer session timeouts.

SSO Session Timeouts for RememberMe can now be configured in the
tokens tab in the realm admin console. Max SSO timeouts for Remember-me
need to be equal or greater than normal max SSO timeouts.

Note that most of the work for this PR was done by @cfsnyder and
is based on his earlier PR https://github.com/keycloak/keycloak/pull/2338.
I just adapted his code to the latest Keycloak API changes.

Signed-off-by: Thomas Darimont <thomas.darimont@gmail.com>

cfsnyder commented:
Awesome @thomasdarimont, thanks for following up and updating that one!

stianst commented on model/jpa/src/main/resources/META-INF/jpa-changelog-2.2.0.xml:
```diff
@@ -57,5 +57,13 @@
 
     </changeSet>
 
+    <changeSet author="keycloak" id="2.2.0.1">
```
Please don't introduce extra changeSet for a specific version, we only want one changeset per-version. Just append to existing changeSet element if there is one.

thomasdarimont replied:
Okay, I moved the changes to the current changelog.

stianst commented on model/jpa/src/main/resources/META-INF/jpa-changelog-2.2.0.xml:
```diff
@@ -57,5 +57,13 @@
+        </addColumn>
+
+        <!-- intialize the SSO_MAX_LIFESPAN_REMEMBER_ME = SSO_MAX_LIFESPAN for all pre-existing realms -->
+        <sql>update REALM set SSO_MAX_LIFESPAN_REMEMBER_ME = SSO_MAX_LIFESPAN where SSO_MAX_LIFESPAN is not null</sql>
```
Would be simpler to allow null and use ssoSessionMaxLifespan if not specified. Having SQL in these files worries me for compatibility reasons (in general, I think this particular SQL is fine though).

stianst reviewed (commented):
Sorry that it took so long to get back to you on this one. Looks OK, but I've added a couple comments inline. We need 'ssoSessionIdleTimeoutRememberMe' as well. This gives more flexibility in how to configure things. For example remember me allows cookie to last 365 days, but still needs a user to use the session at least every 30 days, while a non-remember me session is 30 days and idle 1 day.

thomasdarimont committed:
KEYCLOAK-1267 Add SSO idle timeouts for Remember-Me

We now allow to control the idle timeout for a SSO
session created with "remember me".

Modified the jpa-changelog according to discussions in #3161.

thomasdarimont commented:
FYI I updated the PR with an additional idle timeout for sso remember me. 
Would you mind having a look @stianst and @cfsnyder ?

thomasdarimont commented on model/jpa/src/main/resources/META-INF/jpa-changelog-2.3.0.xml:
```diff
@@ -36,6 +36,16 @@
          </addColumn>
+
+         <addColumn tableName="REALM">
+             <column name="SSO_MAX_LIFESPAN_REMEMBER_ME" type="INT" defaultValueNumeric="3600"/>
```
should the default be 1 hour or 10 hours (as in the examples above e.g.: basicauthrealm.json)?

thomasdarimont commented on model/jpa/src/main/resources/META-INF/jpa-changelog-2.3.0.xml:
```diff
@@ -36,6 +36,16 @@
+
+         <addColumn tableName="REALM">
+             <column name="SSO_MAX_LIFESPAN_REMEMBER_ME" type="INT" defaultValueNumeric="3600"/>
+             <column name="SSO_IDLE_TIMEOUT_REMEMBER_ME" type="INT" defaultValueNumeric="3600"/>
```
I'm not sure what a better default would be here...

thomasdarimont commented on model/jpa/src/main/resources/META-INF/jpa-changelog-2.3.0.xml:
```diff
@@ -36,6 +36,16 @@
          </addColumn>
+
+         <addColumn tableName="REALM">
+             <column name="SSO_MAX_LIFESPAN_REMEMBER_ME" type="INT" defaultValueNumeric="36000"/>
```
set this to 10 hours by default - can anyone think of a better default?

stianst replied:
I don't like having a default for this. Rather values for remember me should be optional and it should use non-remember me values if not specified.

thomasdarimont commented on services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java:
```diff
@@ -97,8 +97,11 @@ public static boolean isSessionValid(RealmModel realm, UserSessionModel userSess
+        boolean sessionMaxOk = (userSession.getStarted() + (userSession.isRememberMe() ? realm.getSsoSessionMaxLifespanRememberMe() : realm.getSsoSessionMaxLifespan())) > currentTime;
+        int maxRememberMeIdleSessionAge = userSession.getLastSessionRefresh() + realm.getSsoSessionIdleTimeoutRememberMe();
+        int maxSsoSessionAge = userSession.getLastSessionRefresh() + realm.getSsoSessionIdleTimeout();
+        boolean sessionIdleOk = (userSession.isRememberMe() && maxRememberMeIdleSessionAge > currentTime) || maxSsoSessionAge > currentTime;
```
Note that  I now take into account the `ssoSessionIdleTimeoutRememberMe`

stianst commented on examples/basic-auth/basicauthrealm.json:
```diff
@@ -6,6 +6,8 @@
     "accessCodeLifespanUserAction": 300,
     "ssoSessionIdleTimeout": 600,
     "ssoSessionMaxLifespan": 36000,
+    "ssoSessionIdleTimeoutRememberMe": 36000,
```
Don't add to existing realm files. That should not be required.

stianst commented on examples/cors/cors-realm.json:
```diff
@@ -6,6 +6,8 @@
     "accessCodeLifespanUserAction": 300,
     "ssoSessionIdleTimeout": 600,
     "ssoSessionMaxLifespan": 36000,
+    "ssoSessionIdleTimeoutRememberMe": 36000,
```
Don't add to existing realm files. That should not be required.

stianst commented on examples/demo-template/testrealm.json:
```diff
@@ -6,6 +6,8 @@
     "accessCodeLifespanUserAction": 300,
     "ssoSessionIdleTimeout": 600,
     "ssoSessionMaxLifespan": 36000,
+    "ssoSessionIdleTimeoutRememberMe": 36000,
```
Don't add to existing realm files. That should not be required.

stianst commented on examples/fuse/testrealm.json:
```diff
@@ -6,6 +6,8 @@
     "accessCodeLifespanUserAction": 300,
     "ssoSessionIdleTimeout": 600,
     "ssoSessionMaxLifespan": 36000,
+    "ssoSessionIdleTimeoutRememberMe": 36000,
```
Don't add to existing realm files. That should not be required.

stianst commented on model/jpa/src/main/resources/META-INF/jpa-changelog-2.2.0.xml:
```diff
@@ -21,7 +21,7 @@
 
         <addColumn tableName="ADMIN_EVENT_ENTITY">
-            <column name="RESOURCE_TYPE" type="VARCHAR(64)"></column>
+            <column name="RESOURCE_TYPE" type="VARCHAR(64)"/>
```
Don't change code that is not relevant to your changes

stianst commented on model/jpa/src/main/resources/META-INF/jpa-changelog-2.3.0.xml:
```diff
@@ -36,6 +36,16 @@
+         </addColumn>
+
+         <!-- intialize the SSO_MAX_LIFESPAN_REMEMBER_ME = SSO_MAX_LIFESPAN for all pre-existing realms -->
+         <sql>update REALM set SSO_MAX_LIFESPAN_REMEMBER_ME = SSO_MAX_LIFESPAN where SSO_MAX_LIFESPAN is not null</sql>
```
This is doomed to fail. See above though. It think this should not be set by default and be optional to specify.

stianst commented on server-spi/src/main/java/org/keycloak/models/utils/RepresentationToModel.java:
```diff
@@ -162,6 +162,12 @@ public static void importRealm(KeycloakSession session, RealmRepresentation rep,
         else newRealm.setSsoSessionMaxLifespan(36000);
+
+        if (rep.getSsoSessionIdleTimeoutRememberMe() != null) newRealm.setSsoSessionIdleTimeoutRememberMe(rep.getSsoSessionIdleTimeoutRememberMe());
+        else newRealm.setSsoSessionIdleTimeoutRememberMe(36000);
```
Same as above. It should not be set by default and use non-remember me values if not specified explicitly.

stianst commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/composites/CompositeRoleTest.java:
```diff
@@ -56,6 +56,8 @@ public void addTestRealms(List<RealmRepresentation> testRealms) {
                 .ssoSessionIdleTimeout(3000)
                 .accessTokenLifespan(10000)
                 .ssoSessionMaxLifespan(10000)
+                .ssoSessionMaxLifespanRememberMe(10000)
```
Shouldn't be required to set this in existing tests. This test has nothing to do with your changes.

stianst commented on testsuite/integration-arquillian/tests/base/src/test/resources/testrealm.json:
```diff
@@ -6,6 +6,7 @@
   "registrationAllowed": true,
   "resetPasswordAllowed": true,
   "editUsernameAllowed" : true,
+  "rememberMe" : false,
```
Why are you changing this?

stianst commented on testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/AbstractConsoleTest.java:
```diff
@@ -77,25 +80,35 @@ public void beforeConsoleTest() {
 
     public void loginToMasterRealmAdminConsoleAs(UserRepresentation user) {
-        loginToAdminConsoleAs(adminConsolePage, loginPage, user);
+        loginToAdminConsoleAs(adminConsolePage, loginPage, user, false);
```
I don't understand the point in adding this. This is not the correct place to test remember me.

stianst commented on testsuite/integration-arquillian/tests/other/mod_auth_mellon/src/test/resources/mellon-realm.json:
```diff
@@ -7,6 +7,8 @@
   "accessTokenLifespanForImplicitFlow" : 900,
   "ssoSessionIdleTimeout" : 2,
   "ssoSessionMaxLifespan" : 36000,
+  "ssoSessionIdleTimeoutRememberMe" : 36000,
```
Please don't add to irrelevant tests

stianst commented on testsuite/integration/src/test/java/org/keycloak/testsuite/adapter/AdapterTest.java:
```diff
@@ -126,6 +126,11 @@ public void testLoginSSOMax() throws Exception {
         testStrategy.testLoginSSOMax();
     }
 
+    @Test
```
Adapter tests are not the correct place to test this. Rather use LoginTest or something else from the base tests inside the new Arquillian testsuite. Please remove all changes to the adapter tests inside testsuite/integration

stianst reviewed (commented):
A few comments inline. In general I think it would be better to use non-remember me values by default and optionally allow setting different values for remember-me.
